Issue 104

JLacoude opened the pull request:
Here is a first version of a solution to the first two problems reported in Issue 104. To handle the `?` operator as the problem is also found when using pure PDO would require to analyze the statement detail and replace the operator with the corresponding function depending on the type of the left and right arguments.

The Rebuilder has been modified as to not require a PDO instance: when a passed value is an array, instead of replacing its placeholder by a csv string of quoted values it creates a csv string of placeholders.

Instead of using regular expression to split the statement it goes through it character by character and when a special one is found it calls a function to handle it. The `Rebuilder->statementPartsHandlers` contains those callables with the character to find as key.

To create Rebuilder for a database vendor you have to implement the RebuilderInterface.

JLacoude committed:
Added test for queries with comments containing :foo

JLacoude committed:
Added test for the postgreSQL shorthand cast operator (::).

JLacoude committed:
Added test for the postgreSQL JSONB operator '?'

JLacoude committed:
Remove the dependency of Rebuilder on ExtendedPdo by changing how array values are handled.

Instead of creating a string from the values we add more parameters to the query and store individual values in the final_values array.

JLacoude committed:
Moved rebuilder tests from AbstractExtendedPdoTest to a new test class RebuilderTest

JLacoude committed:
Removed the __invoke method from the Rebuilder class. Added a RebuilderInterface.

JLacoude committed:
Refactored Rebuilder.php to use a configurable array to parse statements.

Removed all class variables but this array.

JLacoude committed:
Added a PostgreSQL specific rebuilder which decorate the generic one.

JLacoude committed:
Added a test for the Postgres rebuilder to check that it does not accept wrong identifiers.

JLacoude committed:
Implemented the PostgresRebuilder::handleDollarCharacter method.

JLacoude committed:
Added mock PDO object which extends PDO but report 'pgsql' as being its driver

JLacoude committed:
Added a getRebuilder public method to the ExtendPdo class. Currently returns the default rebuilder unless the PDO driver attribute is 'pgsql' in which case it returns the Postgres specific rebuilder.

JLacoude committed:
Replaced call to $handler() with a use of call_user_func($handler) because php 5.3 does not accept this syntax if $handler is not a string.

JLacoude committed:
Replaced null as third argument to mb_substr with the length of the string to get in Rebuilder->prepareNamedPlaceholder because of a bug with php 5.3

JLacoude committed:
Better test coverage.

JLacoude committed:
Add test to check if unclosed multi-line comments are handled.

JLacoude committed:
Add test to check that a / character not followed by a * is not treated like the start of a commented block.

harikt commented on tests/RebuilderTest.php:
```diff
@@ -0,0 +1,196 @@
+    {
+        $stm = "SELECT *
+                FROM pdotest
+                /* Don't bind :foo";
```
seems you missed the closing comment `*/` ?

JLacoude replied:
It is what we want to test: that the Rebuilder does not choke on a statement with unclosed comment.

harikt replied:
Thank you for clarifying.

harikt commented:
Thank you for your good work on this.

One thing I do fear is breaking changes + additional requirements. Eg : mb_string etc.

frederikbosch commented:
@harikt Sounds like a reason to increase the version then: 3.0.

pmjones commented:
This is fantastic! But it does introduce BC breaks, which means it would have to go into a new 3.x branch.

I see you has a "PostgresRebuilder" -- does all the logic in there work with any database, or just Postgred?

pmjones commented:
@JLacoude Actually, let me look at this a little more deeply. Requiring `mb*()` may not be that big a deal; adding the interface is probably a break, but there might be ways to remove the new interface in the interest of maintaining BC.

@harikt Besides the mb-string addition, and the new interface, can you note any other potential BC breaks?

harikt commented:
> Besides the mb-string addition, and the new interface, can you note any other potential BC breaks?

Thank you for the good work on it. To me, this is actually a complex PR. The rebuilder functionality is a complex one and not easy to quickly grasp.

JLacoude commented:
@pmjones 

> I see you has a "PostgresRebuilder" -- does all the logic in there work with any database, or just Postgred?

The PostgresRebuilder decorates the basic Rebuilder to add handling for the _$tag$ a string $tag$_ of writing strings in PostgreSQL. The basic Rebuilder should work with all databases. The goal would be to have one rebuilder per driver as each one has its quirks. Maybe using Traits for the shared parts.

If the BC break is too much, I can drop this PR and make one to correct the 2 bugs from issue 104 which are easily corrected and drop the dependency on PDO from the Rebuilder. And do the mb_string job in another PR for a possible version 3.

JLacoude commented:
> The rebuilder functionality is a complex one and not easy to quickly grasp.

I think part of the problem may come from the parameter passing around all methods. Maybe using a context object would make it easier and less prone to errors.

pmjones commented:
@JLacoude I really like the idea of the new functionality and would prefer to get in in 2.x if at all possible. Let's leave the mb-string stuff in place for now, but if you can reduce the complexity (or at least make easier to grasp) via the Context object you suggest, that would really help with review.

Thanks a ton for your work here.

JLacoude committed:
Modified the rebuilder to use a RebuilderState object to pass around handlers.

frederikbosch commented:
@JLacoude Since #109 was merged, this will conflict your PR. That PR throws an exception when there is a missing parameter when using the numbered placeholders (question mark).

JLacoude committed:
Merged Issue-109.
Add LegacyRebuilder.
Remove automatic selection of Rebuilder. Replace by a setRebuilder method in ExtendedPDO

JLacoude commented:
As the changes seem to feel hazardous, I added a LegacyRebuilder. It is the default one and mostly does what the old rebuilder did: use regexp to extract parts of the queries. I just kept the decoupling from PDO when dealing with arrays.

To let users chose the Rebuilder they want, I added a setRebuilder method to ExtendedPDO.

JLacoude committed:
Merged Issue-109.
Add LegacyRebuilder.
Remove automatic selection of Rebuilder. Replace by a setRebuilder method in ExtendedPDO

frederikbosch commented:
@JLacoude I think that is a nice solution!

harikt commented on src/ExtendedPdo.php:
```diff
@@ -992,6 +992,23 @@ public function prepareWithValues($statement, array $values = array())
+    {
+        $oldRebuilder = $this->rebuilder;
+        $this->rebuilder = $rebuilder;
+        return $oldRebuilder;
```
I don't think you need to return the old builder. What is the purpose of returning it?

JLacoude replied:
It's so you can easily swap it for some operations without needing a getter.

``` php
<?php
class Example{
    public function test(ExtendedPdoInterface $db){
        $original = $db->setRebuilder(new MockRebuilder());
        /* some operations */
        $db->setRebuilder($original);
    }
}
```

pmjones commented:
@JLacoude I have started a 3.x branch for this package, so BC breaks are allowed. If you are still willing, rebase and resend a PR to that branch, and we can merge it. Thanks for your efforts, and your patience.

JLacoude commented:
@pmjones Still willing to. But as BC breaks are allowed, would allowing back the use of multiple queries at once be a good idea?

pmjones commented:
> would allowing back the use of multiple queries at once be a good idea?

Might as well, since it's a BC break. Thanks for your help!

pmjones commented:
@JLacoude Any progress on this? I know it's a lot of work; thanks for any effort you can still provide.

JLacoude commented:
@pmjones Work and other projects have been taking a lot of my free time this year so I could not progress on this like I would I appreciated. I'll try to get something around the end of September.

pmjones commented:
@JLacoude No problem, and thanks for any efforts you can provide.

pmjones commented:
@JLacoude please see #134 as a replacement for this PR; all credit to you has been retained. Let me know if it looks OK. Thanks for all your good work here!

pmjones commented:
This is obviated by #138 which has been merged.
